### PR TITLE
Optimize sharded bcast_h op

### PR DESCRIPTION
**Pull request by johanna-rock-tt**
### Ticket
- [Link to Github Issue.](https://github.com/tenstorrent/tt-metal/issues/9472)

### Problem description
- bcast for h dim implementation was not optimal and thus slow

### What's changed
Performance optimization for sharded bcast op in the h dim. It re-uses in1 for all h tiles of in0 per core instead of re-loading in1 for each of them. The interleaved implementation is unchanged.

C++ tests pipeline [PASSING]: https://github.com/tenstorrent/tt-metal/actions/runs/9548312511

### CODE REVIEW

**Review by shwetankTT — APPROVED** [review]
looks good to me.

File: tt_eager/tt_dnn/op_library/bcast/kernels/dataflow/reader_bcast_h_sharded.cpp [hunk 1]
```diff
@@ -5,13 +5,16 @@
 #include <stdint.h>
 #include "dataflow_api.h"
 
+#include "debug/dprint.h"
```
Comment (TT-BrianLiu) [inline tt_eager/tt_dnn/op_library/bcast/kernels/dataflow/reader_bcast_h_sharded.cpp:8]: remove

**Review by TT-BrianLiu — APPROVED** [review]
Minor comments

**Review by tt-aho — APPROVED** [review]
lgtm after removing the debug prints.

This was a somewhat known perf issue/todo to fix for how in1 is used to minimize reads. Need to eventually propagate for the other bcast flavors 😞

**Comment by johanna-rock-tt** [comment]
Speedup depends on the input shape (in0 h tiles per core) and is different for block and width sharding.
Falcon 40b prefill shapes [S, 8192] as an example:

```
block sharded S=256
1.12 x
block sharded S=2048
4.8 x

width sharded S=32
1.3 x
width sharded S=2048
6.7 x
```

**Comment by johanna-rock-tt** [comment]
I think interleaved might benefit even more from this optimization. It starts off 10x slower (L1 interleaved in0) than the block sharded variant and I think a lot of that comes from how we read in1 there.

I'll track that in an issue, I think there's also further optimizations we can do for the sharded variant (mcast in1 in block sharding, or the RM wrap around a tile trick that we also do in the layernorm kernel). Or is there already a ticket that tracks this?

**Comment by tt-aho** [comment]
I did a search yesterday but didn't find any issues for this. Might have been something that was brought up at one point but wasn't noted down

**Pull request merged**